### PR TITLE
Split CST and derivings (with an OCaml splitter)

### PR DESCRIPTION
**Pull request by Niols**
Same as #126 except the splitter has been rewritten in OCaml. This does not make the compilation faster or clearer. It does allow for a better documentation of the process (in particular, the regexp can be split and documented in tiny bits). More importantly, it fixes the compilation on Windows, which we will see once we integrate the new CI (see #127).

We have to choose between this PR, #126 or none of the two if we don't validate the concept. @yurug, WDYT?

### CODE REVIEW

File: src/CST_derivings_generator.ml [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  in
+  read ();
+  close_in ichan;
+  Buffer.contents out
+
+(* Add type equalities to all type definitions. *)
+
+let type_foo_eq_regexp =           (* this regular expression matches: *)
+  "\\(type\\|and\\)\\([^=\n]* \\)" (* "type" or "and" followed by anything that ends by a space *)
+  ^ "\\([^ =\n]+\\) *="            (* anything that looks like a type identifier, followed by = *)
```
Comment (yurug) [inline src/CST_derivings_generator.ml:117]: I see that you have answered my previous comment here by under specifying the language for type identifier. This is OK and leads to clearer code in my opinion.

File: src/CST_derivings_generator.ml [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  ^ "\\([^=\n]*=\\|\\) *"          (* possibly some other stuff followed by equal (type equality) *)
+  ^ "\\([^ a-z\n][^=\n]*\\|\\)"    (* anything that does not start with a lowercase character *)
+  ^ "$"                            (* because that would also be a type equality *)
+let type_foo_eq_regexp = Str.regexp type_foo_eq_regexp
+(* Note: there can be type equalities that do not syntactically start with a
+   lowercase character (eg. when there is a module name involved), but the
+   current regexp is good enough for our use and our definition of the CST. *)
+
+let type_foo_eq_repl = {|\1\2\3 = \2CST.\3 = \5|}
+  (* the replacement is of the form "type foo = CST.foo = ..." *)
```
Comment (yurug) [inline src/CST_derivings_generator.ml:127]:
```suggestion
  (* The replacement is of the form "type foo = CST.foo = ..." *)
```

**Review by yurug — APPROVED** [review]
I did some research and I found no cleaner way to achieve that.

At some point, I was wondering if using `cpp` with appropriate `#ifdef` could not do the trick more efficiency but that's not clear.

Anyway, I am glad that this issue is fixed. Thank you very much Nicolas!

File: src/CST_derivings_generator.ml (hunk 1 above)
Comment (Niols) [reply src/CST_derivings_generator.ml:117]: I also think this is a better way to do things indeed. But we can bring this regexp back to Shell if we decide we prefer the Shell approach of #126 over the OCaml approach of this PR.

File: src/CST_derivings_generator.ml (hunk 2 above)
Comment (Niols) [reply src/CST_derivings_generator.ml]: Done in 37f77ac!

**Comment by Niols** [comment]
The other idea that I had in the meantime was to write a tiny custom PPX just for the compilation of this CST file. This would avoid a textual manipulation of the type definitions and make this part of the code more robust. I can give it a try and open yet another PR. We'd have to choose between the approaches:

1. Textual manipulations by a Shell script (#126)
2. Textual manipulations by an OCaml executable (this PR #128)
3. [Parsetree](https://ocaml.org/api/compilerlibref/Parsetree.html) manipulations by an OCaml executable (*future work*)
4. *Any other idea.*

What do you think would be preferable? I would say 2. is better than 1., but maybe 3. is better than both and I should just do that?

**Comment by yurug** [comment]
@Niols Can you please rebase your branch on top of master to check that the new CI is green?

Gitlab offers a `rebase` button but Github, unfortunately, does not.

**Comment by Niols** [comment]
Just rebased master into this PR. I had already checked the CI so it should be all green unless I did something very wrong!

**Comment by Niols** [comment]
All green!

**Pull request merged**